pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -83,8 +91,61 @@ func (s *preFilterState) Clone() framework.StateData {
 // failed by this plugin schedulable.
 func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
+	}
+}
+
+// isSchedulableAfterNodeChange is invoked whenever a node changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodeAffinity) isSchedulableAfterNodeChange(pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		pl.logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	requiredNodeAffinity := nodeaffinity.GetRequiredNodeAffinity(pod)
+	if originalNode == nil {
+		// a new node creation event.
+		// We check if the node matches with a node affinity and queue this Pod if match.
+		matched, err := requiredNodeAffinity.Match(modifiedNode)
+		if err != nil {
+			pl.logger.Error(err, "failed to match pod to node", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+			return framework.QueueAfterBackoff
+		}
+		if !matched {
+			pl.logger.V(4).Info("node was created but doesn't match Pod's NodeAffinity", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+			return framework.QueueSkip
+		}
+		pl.logger.V(4).Info("node, which matches with the afifnity, was created", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueAfterBackoff
 	}
+
+	// Modifications may or may not be relevant. If the entire
+	// labels is as before, then something else must have changed
+	// and we don't care.
+	if apiequality.Semantic.DeepEqual(originalNode.Labels, modifiedNode.Labels) {
```

Comment:
How about like this for the modification event? 
Your current code return QueueAfterBackoff every time an unrelated label on the Node, which already matches with the Pod's nodeaffinity, is updated. But, ↓ returns QueueAfterBackoff only when the change makes the Node match with the Pod's nodeaffinity

```go
// pseudo code

wasMatched, err := requiredNodeAffinity.Match(originalNode)
isMatched, err := requiredNodeAffinity.Match(modifiedNode)

if !wasMatched && isMatched {
  // This modification make this Node matched with Pod's NodeAffinity. 
  return framework.QueueAfterBackoff
}

return framework.QueueSkip
```

Reply:
@sanposhiho updated.